### ISSUE TITLE
Comparison between EQP solvers

### ISSUE THREAD

**Issue by antonior92** [antonior92/ip-nonlinear-solver#2]
In the last two days I have been busy installing CUTEst and pycutestmgr (which took way longer than I was expecting) and performing tests. I compared the direct factorization method with two approaches using projected CG:
- One using the normal equation approach;
- The other using the augmented system approach;

I performed experiments on the following set of problems:
``CVXQP1_S, CVXQP2_S, CVXQP3_S, CVXQP1_M, CVXQP2_M, CVXQP3_M, CVXQP1_L, CVXQP2_L, CVXQP3_L, CONT-050, CONT-100, DPKL01, MOSARQP1, DUAL1, DUAL2, DUAL3, DUAL4, PRIMAL1, PRIMAL2, LASER``.The description of the problems can be found in \[1\] (convex QP problems). Each problem will be represented by a single circle on the following plots.

I compared the methods regarding the execution time and the error of optimality measures. I used CG methods with very low tolerances.

About the execution time: The time it takes to factorize the augmented system is much smaller than the time it takes to solve the system by direct factorization of the KKT matrix (About 1000x smaller for large problems) as illustrated in the graph bellow. 
![screenshot 2017-06-02 15 40 11](https://cloud.githubusercontent.com/assets/16557411/26739887/e8c45f3c-47a9-11e7-8b4e-b4433a8c6ef9.png)

Furthermore, the time to solve the problem using any of the CG approaches is usually much smaller than the time to solve them by using direct factorization.
![screenshot 2017-06-02 15 40 25](https://cloud.githubusercontent.com/assets/16557411/26740031/65b0460a-47aa-11e7-9cb1-8a665effcfdf.png)

The concern about large roundoff errors when using the CG approach is not unfounded. The next scatter plot shows that for some of the large problems both the error in the equality constraint
and the Lagrangian gradient norm are very large when they should be zero. 

![screenshot 2017-06-02 15 39 57](https://cloud.githubusercontent.com/assets/16557411/26739916/fe37fab8-47a9-11e7-8086-299f8d6a14fd.png)

References
------------
\[1\] [Istvan Maros and Csaba Meszaros  "A repository of convex quadratic programming problems", 
Optimization Methods and Software (OMS) Volumes 11&12 (December, 1999), 671-681](http://www.doc.ic.ac.uk/rr2000/DTR97-6.pdf)

**Comment by antonior92** [comment]
The **augmented system approach**(using LU factorization) is a little slower (but no more than 2x slower) than the **normal equation approach**. Nevertheless I think the convenience of using something that already is in scipy may justify its use

**Comment by antonior92** [comment]
Something have to be done regarding the large errors when using projected CG solver (that results in large values of errors in the equality constraints and in the Lagrangian gradient norm), otherwise this could compromise the algorithm execution. I think I should implement some of the refinements described in \[2\] in order to see if I can get some improvement.

\[2\]&nbsp;&nbsp;&nbsp; [Gould, Nicholas IM, Mary E. Hribar, and Jorge Nocedal. "On the solution of equality constrained quadratic programming problems arising in optimization." SIAM Journal on Scientific Computing 23.4 (2001): 1376-1395.][2]

[2]: https://pdfs.semanticscholar.org/d959/8c935921efb5c4459fd6c55cf501cd578b45.pdf

**Comment by nmayorov** [comment]
Amazing job.

My conclusion from the report:
- Augmented system approach can be chosen as the default method. Whether to support another method is up to you, I guess it is not very necessary as it is not an established practice to rely on external code in scipy.
- It looks useful or even necessary to implement "iterative refinement" to reduce numerical errors.

Anything else you from you? 

Can you post some version of the code used for comparison? Or maybe describe briefly how you did it (i.e. setup CUTEst)? 

-------------------


Another branch of work --- consider small problems and dense solvers.

**Comment by antonior92** [comment]
There is a few available options to install CUTEst (or its predecessor CUTEr) for python:
1. Install a library called [pyOPUS](http://fides.fe.uni-lj.si/pyopus/) that have a wrapper for CUTEr.
2. Use a library called [CUTEST.py](https://github.com/kenjydem/CUTEST.py)
3. Use the solution provided by Prof. Árpád Bűrmen in his webpage [link](http://fides.fe.uni-lj.si/~arpadb/software-pycuter.html)

I would not recommend the first two approaches because both ``pyOPUS`` and ``CUTEST.py`` have several dependencies. I tried both and didn't manage getting them running and lost a lot of time. The solution provided by Prof. Árpád Bűrmen is much more clean, but I was having some troubles with it as well.

Fortunately I remember that Felix Lenders used CUTEst in his Pull Requests and I reached him out. He just provided me some modified files he have implemented based on Prof. Árpád Bűrmen  solution.  And this was what worked out for me.

**Comment by antonior92** [comment]
I describe how to install it here: 
https://gist.github.com/antonior92/9fd6ac9f44c164ee42555e3a7dc5a332

Using these two files that Felix provided to me:
[pycutestmgr.py.txt](https://github.com/antonior92/ip-nonlinear-solver/files/1048823/pycutestmgr.py.txt)
[pycutestitf.c.txt](https://github.com/antonior92/ip-nonlinear-solver/files/1048824/pycutestitf.c.txt)

**Comment by antonior92** [comment]
The code I used for the comparison:
https://gist.github.com/antonior92/f8e6d1693010eea286887e0688a62ce8

**Comment by antonior92** [comment]
I have already implemented some small problems on my test [files](https://github.com/antonior92/ip-nonlinear-solver/blob/master/ipsolver/test/test_projected_cg.py). Nevertheless I think I need to do more tests. Because I need to be sure nothing is wrong with my logic before moving on to the iterative refinement.

**Comment by mdhaber** [comment]
This looks great, Antonio.

It seems that Nikolay is being careful to keep up with the theory. This is
great (Thanks, Nikolay!). I am probably not going to do that unless you
have trouble with it. So sometimes I may ask questions at the risk that are
uneducated and don't make sense. When that's the case, just tell me, or try
to explain what I am missing : )  But I hope that more often than not I am
able to help nonetheless.

1) I've implemented a CG method for solving a system of linear equations
before. I'm assuming that even with the variant you're using to account for
the constraints, the procedure is iterative, so the error depends on your
iteration stopping criteria. Is it simple to explain why you're considering
round-off error as fixed for each of the CG approaches, that is, why you
cannot reduce these norms with tighter stopping criteria?

2) Are the large norms a problem? Considering that this is a sub-step in
your overall algorithm, small errors might not be... but these seem quite
large, so I'm not sure.

**Comment by antonior92** [comment]
Your questions are very good Matt.  What you are saying is true, the procedure is iterative, so as you iterate you get progressively better results. There is however a limit. In the case of a linear system with size ``n``after ``n`` interaction you get (in theory) the exact result. For the case of the EQP problem with ``n``variables and ``m``constraints after ``n-m`` iteration you get the exact result (in theory). So when I put a very small tolerance as I did (``1e-20``) I expect the result to be very close to the actual one. If it is not there is something wrong.

And yes, during the algorithm execution I will actually set a very loose tolerance, and, in this case, it is fine to have large errors, because I was only interested on an approximate result. However, during this test, I set the problem with a very tight tolerance and large errors are not acceptable in this case. 
Because if large error occurs when I set tight tolerances, then the tolerance doesn't have much significance.

Therefore, what I intend to do is: first to check my logic and see if nothing is wrong. Second, if I sure that nothing is wrong, I will implement the refinement described in \[2\].


\[2\]&nbsp;&nbsp;&nbsp; [Gould, Nicholas IM, Mary E. Hribar, and Jorge Nocedal. "On the solution of equality constrained quadratic programming problems arising in optimization." SIAM Journal on Scientific Computing 23.4 (2001): 1376-1395.][2]

[2]: https://pdfs.semanticscholar.org/d959/8c935921efb5c4459fd6c55cf501cd578b45.pdf

**Comment by mdhaber** [comment]
I think I see, so let me rephrase so you can confirm:
You are concerned that there might be something wrong with your
implementation of the algorithm, because even though you set tight
tolerances, the error is high.
The other possibility you are considering is that even though the
implementation is correct, the theoretical result is not achieved because
of roundoff error. (If this is the case, you might be able to confirm this
by modifying your code to use arbitrary precision arithmetic using SymPy or
something, but this might be more trouble than it's worth.)
The third possibility I can think of is that the theory does not actually
guarantee convergence (even in the absence of roundoff error) in some
cases... but it sounds like you've ruled that out.

So... if the tolerance was not achieved, what caused the algorithm to
terminate? Do you have it stop if it reaches n-m iterations even if the
tolerance is not met? What if you allowed it to continue? What if you
stopped it early?
It might be informative to play with a problem that gave really poor
results and plot the norms as a function of the iteration number.

**Comment by mdhaber** [comment]
Also I want to clarify that by :
"(If this is the case, you might be able to confirm this by modifying your
code to use arbitrary precision arithmetic using SymPy or something, but
this might be more trouble than it's worth.)"
I certainly don't mean that arbitrary precision arithmetic should be used
in the final version of the code. I mean that it can be used to help you
confirm whether roundoff error is the culprit.

**Comment by antonior92** [comment]
Hey Matt, the roundoff error is a reported problem of this mathod. In \[2\] they provide a solution to this problem that I think I could try.  I dont think we would need to use arbitrary precision.

The possibility of errors in the implementation cannot be rulled out and I want to be careful not to atribute other problems to this known large roundoff error

I will plot the error throught the iteractions, it might give me a better idea of what is happening.

**Comment by antonior92** [comment]
I will do some tests today. I will get back to this soon with something more substantial

**Comment by antonior92** [comment]
The errors are indeed the roundoff errors. In order to understand why, it is important to understand something about the projected CG method. Consider the EQP problem ``min 1/2 x.T G x + c.T x`` subject to ``A x = b``

The CG method choose an initial point  ``x0`` such that ``A x_0 = b``. And all the updates ``p``
are chosen such that ``A p = 0``, in order words, the updates belongs to the null space of A.
And for exact arithmetics this works fine and we get the optimal value of  ``min 1/2 x.T G x + c.T x`` 
while satisfying the constraints.

The point is that for ill-conditioned problems ``A p != 0``, and we start distancing ourselves along the iterations. I did a test with the problem ``CVXQP3_M`` and came to the conclusion that the norm
``||A x_0 = b||`` is indeed increasing along the iterations:
![error_per_iteration](https://cloud.githubusercontent.com/assets/16557411/26786243/51d76094-49dc-11e7-903f-df462bd4849b.png)

This is a clear indicator that the problem is indeed the roundoff errors.

**Comment by antonior92** [comment]
Since the roundoff errors are indeed the problem I will take some time to read \[2\] carefully and implement whatever they have done to attenuate the problem.



\[2\]&nbsp;&nbsp;&nbsp; [Gould, Nicholas IM, Mary E. Hribar, and Jorge Nocedal. "On the solution of equality constrained quadratic programming problems arising in optimization." SIAM Journal on Scientific Computing 23.4 (2001): 1376-1395.][2]

[2]: https://pdfs.semanticscholar.org/d959/8c935921efb5c4459fd6c55cf501cd578b45.pdf

**Comment by nmayorov** [comment]
> The point is that for ill-conditioned problems A p != 0

Can you elaborate on that?

-------------------

Overall, can you refine your plan for the following week? Are you going to consider dense problems before moving to the main solver?

**Comment by antonior92** [comment]
> The point is that for ill-conditioned problems A p != 0

The point is that in exact arithmetics I should have ``A p = 0``, where ``p`` is the step. However, when dealing with floating point arithmetics, I have an error. That way, when  we should have 
``A(x+p) = Ax + Ap = b+0 =b`` we get instead ``A(x+p) = Ax + Ap = b + error``. And the point I tried to show with the graph is that this error is raising through the iterations.

**Comment by antonior92** [comment]
And Nikolay, I intend to move to the main solver as soon as I get everything working for sparse matrices.

I don't think a solver for dense EQP problems is much relevant right now because the matrices
``A`` and ``G`` that arises during the trust-region problem are naturally sparse (with lots of zeros) even if all the user provided matrix are dense.

I think we should discuss this further, but I think we should prioritize sparse matrices. I think this way because many nonlinear programing problems that arises are naturaly sparse. Many nonlinear solvers doesn't even provide the option of the user provided matrix being dense. For instance, IPOPT solver only allows sparse matrices [documentation](https://www.coin-or.org/Ipopt/documentation/node23.html#SECTION00053180000000000000) also does the Julia interface for nonlinear programing (juMP)[http://jump.readthedocs.io/en/latest/] and many other solvers.

Scipy already have a SQP method that works for dense matrices, so I think the ability to deal with large-scale problems is more pressing. Furthermore, the KNITRO method (the one I intend to implement) was conceived for large-scale problems, where the matrices are mostly sparse. I don't think it will be a good choice for dense problems no matter what.

**Comment by nmayorov** [comment]
> The point is that in exact arithmetics I should have A p = 0, where p is the step. 

My question was more about why *ill-conditioned* matrices cause such degradation, i.e. what determines the growth of the numerical error. Maybe you have developed some intuition?

**Comment by antonior92** [comment]
I am reading the paper I have previously mentioned and they actually provide a numerical expression for this error. It grows with the squared condition number of the matrix (for the case of the normal equation approach).  So there is indeed a direct relation

**Comment by mdhaber** [comment]
Great, then it sounds like the problem you are experiencing is not unusual and there is reasonable hope that the suggestions in the paper will fix it.

If they do, you will be done with all the EQP solver approaches mentioned in your proposal (I'm assuming null space from the proposal is the "direct factorization" method above), and will be ready to move on to the Trust Region IP Method?

In general, I think it is best to get proof of concept of everything working before stopping to "make better". However, it sounds like these improvements to the CG method to abate the round-off error are important enough to stop for.  It could be frustrating to build the rest on top of an unstable EQP routine.  But I agree that you should not stop to work on a dense EQP solver.

**Comment by nmayorov** [comment]
> I don't think a solver for dense EQP problems is much relevant right now because the matrices
A and G that arises during the trust-region problem are naturally sparse (with lots of zeros) even if all the user provided matrix are dense.

Is it because the slack variables (used for adding the barrier) have diagonal Jacobian matrix? Nevertheless, It seems to me that when the number of inequality constraints is small and the original probmel is "dense" then `G` will be quite dense as well. Am I correct here?

> I think we should discuss this further, but I think we should prioritize sparse matrices. I think this way because many nonlinear programing problems that arises are naturaly sparse. Many nonlinear solvers doesn't even provide the option of the user provided matrix being dense. For instance, IPOPT solver only allows sparse matrices documentation also does the Julia interface for nonlinear programing (juMP)[http://jump.readthedocs.io/en/latest/] and many other solvers.

I agree about prioritizing, but not about that we should do the same as others no matter what. I think there should be a good reason for "limiting" our procedure like this. Basically it comes down to whether considering sparse/dense problems separately will give significant time savings. And I don't think that we should require any "sparseness" in user-provided matrices. Does it make sense?

> I don't think it will be a good choice for dense problems no matter what. 

Could please explain this breifly?

-----

My conclusion so far is that you should move on to the main solver after implementing this "refinment" procedure, but later (if nothing prevents that and it makes sense, see the above point) implement the dense solver as well. As for the implementation, it seems that you should simply replace your linear operators with matrices computed with QR decomposition (which explained sufficiently in Num. Opt.) Are you generally agree or I missed something important?

**Comment by antonior92** [comment]
> I don't think it will be a good choice for dense problems no matter what. 

I may have exaggerated things a little bit. But my point is that the algorithm was conceived thinking of large-scale problems. That way, it will probably be outperformed for dense small/medium size problems for algorithms made specifically for this kind of problem.

Much like my ``trust-exact`` method outperforms the ``trust-ncg`` trust-region method for medium/ small problems (If you remember my pull request).

> Is it because the slack variables (used for adding the barrier) have diagonal Jacobian matrix? Nevertheless, It seems to me that when the number of inequality constraints is small and the original probmel is "dense" then `G` will be quite dense as well. Am I correct here?

You have a point. If there are few inequality constraints the problem will be quite dense. On the other hand, if there are lots of inequality constraints it will contain a lot of zeros.

> My conclusion so far is that you should move on to the main solver after implementing this "refinment" procedure, but later (if nothing prevents that and it makes sense, see the above point) implement the dense solver as well. As for the implementation, it seems that you should simply replace your linear operators with matrices computed with QR decomposition (which explained sufficiently in Num. Opt.) Are you generally agree or I missed something important?

It make sense to me. I could modify my function ``projections`` easily to deal with dense matrices. I would still be using Conjugate Gradient method in order to deal with the equality constraints, but it may improve performance for completely dense matrices.

I don't see it as a priory, but I can include it in this week activities. What do you think?

**Comment by nmayorov** [comment]
> If you remember my pull request

I sure do 😃 

> You have a point. If there are few inequality constraints the problem will be quite dense. On the other hand, if there are lots of inequality constraints it will contain a lot of zeros.

Looks like a good situation to consider both cases in my opinion.

> I don't see it as a priory, but I can include it in this week activities. Or maybe I should leave it for latter. What do you two think about it?

My answer --- up to you. But it seems like a rather simple problem, isn't it? So I guess you can fit it into this week and then mentally move on to the main solver.

**Comment by antonior92** [comment]
You have a point. I will do it this week then, while it is fresh in my mind. I will update the wiki.

**Comment by mdhaber** [comment]
Isn't the tricky thing to decide when to use the dense method?
Easy if it's just an option.
Tougher if you have to benchmark to determine which methods do use.

**Comment by antonior92** [comment]
I think the dense method would be better only if there are few inequality constraints (besides the obvious fact that the Jacobian and Hessian should be dense).

**Comment by antonior92** [comment]
The modification seems to have worked out. Look at the same test I did before:
![error_per_iteration_corrected](https://cloud.githubusercontent.com/assets/16557411/26832169/fdccfdaa-4aa4-11e7-835d-d8fbccd2b22d.png)

But I think I accidentally introduced something that is making my program to crashes when I try to run the very large problems from CUTEst. I will just figure out what is wrong and come back with more results soon.

**Comment by mdhaber** [comment]
That plot looks amazing! Is it much slower? (doesn't matter, just curious)

**Comment by antonior92** [comment]
The time per interaction is more or less the same. But the new implementation goes through more interaction, because some interactions were prematurely interrupted because of the roundoff errors in the previous implementation.

However, there is something crashing (everything just stop working) when I run on very large problems.  And it happens only for large problems. I will check everything to try to figure out what
it is. And I hope to come back with a fix soon. Maybe I did something wrong with memory, or maybe it is a problem with the way I am using pycutest cache folder... I need some time to figure it out.

**Comment by antonior92** [comment]
Ok! Just figure it out! The problem was on my implementation of some orthogonality measure function. Just fixed it.

**Comment by mdhaber** [comment]
W00t! Great work : )

**Comment by antonior92** [comment]
I think the projected CG is working fine now. So I will summarise my findings.

-------------

I did test on a few of convex QP problems from CUTEst.  Each different problem is represented by a dot on the bellow graph. I am comparing the optimality measure ``1/2 x.T G x + c.T x + f`` in the y-axis
with the fulfilment of the constraints ``\|\| A x - b\|\|``. This is a little different from the plot I did first but I think it make more sense.

Before the modification I had a very large constraint violation for some of the problems, as can be seen on the graph b
![optimality_x_error_before](https://user-images.githubusercontent.com/16557411/26845487-f4f9edfa-4acc-11e7-81f7-33e7558d1357.png)

After the modification the constrains seems to be fulfilled with the desired level of accuracy:
![optimality_x_error_after](https://user-images.githubusercontent.com/16557411/26845492-f937e174-4acc-11e7-91a6-97f750355436.png)

----

As I have explained before this is due to roundoff errors that accumulate during the iterations.
I plot the constraint violation for the problem ``CVXQP3_M`` before and after the modification.

Before:
![error_per_iteration](https://user-images.githubusercontent.com/16557411/26845806-169b833c-4ace-11e7-8f08-bd125d4966f9.png)

After:
![error_per_iteration_corrected](https://user-images.githubusercontent.com/16557411/26845823-22141c42-4ace-11e7-9267-e8cfd46d0d0f.png)

-------------------
About the execution time, things have not changed significantly from before: The normal equation approach is slightly (no more than 2x) faster than he augmented system approach. The Direct factorization is significantly slower (more than 10x) for very large problems and can be slightly faster for small problems compared to both CG methods.


![execution_time_after](https://user-images.githubusercontent.com/16557411/26845501-004687a4-4acd-11e7-9ff9-48bb6aa0a607.png)

**Comment by nmayorov** [comment]
> The Direct factorization is significantly slower (more than 10x) for very large problems and can be slightly faster for small problems compared to both CG methods.

I suppose it is not worth it to include direct factorization because scipy is not designed to solve small problems very fast (i.e. we are not going to solve many of them real time or anything like that). Not that you wanted to do that, just my justification why we probably should not bother here.

---------- 

Overall great job so far. Will you implement the alternative QR factorization approach for dense problems? How about doing it tomorrow and making the blog post tomorrow as well? The blog post is not strictly necessary I suppose, but otherwise you will have to do it somewhat later this week (probably on another topic).

------------

As for the code goes: I suppose line-by-line review will be made (I don't mind doing it, if Matt is not very eager to) after you send some batches to scipy in PR. I don't know how to schedule it: I guess it is better not to postpone it to the very end of the summer, but on the other hand it should be convenient to you, i.e. don't split the work for the sake of splitting. What do you think about it? When do you think it will make sense to send some code to scipy repo?

**Comment by antonior92** [comment]
So, I still have to do three things:

- [ ] Add capability to deal with trust-region constraints
- [ ] Implement projections for dense matrix;
- [ ] Write blog post about the method;

I intend to do the three of them in the next one or two days so I can start to implement the main solver still this week.

About the PR request: I intend to send a PR with the main solver probably in 3 or 4 weeks. I don't see the point on sending a PR with the projected CG, since I see it mostly as a substep of the solver.

What do you two think about this plan?

**Comment by nmayorov** [comment]
It all sounds good to me.

**Comment by antonior92** [comment]
Ok, so I think I am pretty much done with the implementation of the CG method. I included the capability to deal with dense matrices. And also the possibility of including trust region constraints. Furthermore I added a bunch of tests verifying diverse properties of the method.

**Comment by nmayorov** [comment]
@antonior92 it all looks quite solid indeed. 

One thing which is a bit strange to me: wouldn't it be better if `projected_cg` takes the matrices of the original problem and then computes required operators inside? This way it will be the only function a user needs to understand to solve his EQP problem. What was your motivation to split it like this?

**Comment by antonior92** [comment]
Hey Nikolay,

I was thinking about the main solver when I implemented it this way. The idea is that some factorizations needed by the ``projected_cg``method are also needed on other substeps of the main solver. So this way it is implemented it is easy to reuse the linear operators.

**Comment by antonior92** [comment]
Actually, I think I have already covered all the factorisations I will need for the main solver on this function ``projections``.  The operator ``LS`` is not actually needed by the ``projected_cg`` method, but I already implemented it, thinking about the main solver.

**Comment by nmayorov** [comment]
Yeah, it makes sense. If we decide to make EQP solver as a public function, then you can make a self-contained version as well.

---------------

Do you have anything specific to ask from me or Matt at the moment? Like review or check some things.

**Comment by antonior92** [comment]
I just created the blog post about the projected CG method. https://antonior92.github.io/posts/2017/05/projected-CG/

I don't think I need anything specific from you two right now. The daily discussions and some comments on the code, as you two are already doing, are serving its purpose and I think things are progressing well enough.

**Comment by antonior92** [comment]
I intend to do some final revisions today on the projected CG.

My plan for tomorrow was to do a theoretical revision about the main solver and present a implementation plan. What do you two think about this?

**Comment by mdhaber** [comment]
I think it sounds good, but what sort of theoretical revisions do you need?

**Comment by nmayorov** [comment]
>  I just created the blog post about the projected CG method. https://antonior92.github.io/posts/2017/05/projected-CG/

Will read carefully.

> I think things are progressing well enough.

I agree with that.

> My plan for tomorrow was to do a theoretical revision about the main solver and present a implementation plan for it. What do you two think about this schedule?

I don't mind at all, writing a good plan (in wiki) is an important step to work efficiently afterwards. But what are your thoughts overall about being somewhat behind on the original plan? I don't think it is a problem at all, I believe the progress is good.

**Comment by antonior92** [comment]
> I think it sounds good, but what sort of theoretical revisions do you need?

I want to read the main paper describing the method carefully thinking about how to proceed about each step in order to write a more careful implementation plan.

I think having a good implementation plan will definitely worth the extra time I will spend doing it.

> But what are your thoughts overall about being somewhat behind on the original plan?

When I wrote my GSoC proposal I did it accordingly to the best of my knowledge at the time. That said, there are some thing I have not consider, for instance the need to correct to implement roundoff errors on projected CG method, which took me some time this week.

On the other hand, I also included time to do things I will not actually need to do. For instance, I will not have to implement preconditioners (for which I have schedule a whole week) since the trust-region method uses some variable change internally. Other example, the access to CUTEst will probably save a lot of time in benchmarking the code since I will not have to actually implementing the benchmarks (for which I have schedule 1 or 2 weeks).

So maybe we are behind the original plan, for which I have only allocated one week for projected CG. But, looking at the big picture, we are ok.

**Comment by mdhaber** [comment]
Like you and Nikolay, I agree that while you are not on the schedule you
wrote, you're certainly not behind. Great work!
